src/backend/cpu/convolve.hpp

```diff
@@ -12,12 +12,29 @@
 
 namespace cpu {
 
-template<typename T, typename accT, int baseDim, bool expand>
-Array<T> convolve(Array<T> const& signal, Array<accT> const& filter,
+template <typename T, typename accT, int baseDim, bool expand>
+Array<T> convolve(Array<T> const &signal, Array<accT> const &filter,
                   AF_BATCH_KIND kind);
 
-template<typename T, typename accT, bool expand>
-Array<T> convolve2(Array<T> const& signal, Array<accT> const& c_filter,
-                   Array<accT> const& r_filter);
+template <typename T, typename accT, bool expand>
+Array<T> convolve2(Array<T> const &signal, Array<accT> const &c_filter,
+                   Array<accT> const &r_filter);
 
-}  // namespace cpu
+template <typename T, typename accT>
+Array<T> convolve2(Array<T> const &signal, Array<accT> const &filter,
+                   const dim4 stride, const dim4 padding, const dim4 dilation);
+
+template <typename T, typename accT>
+Array<T> conv2DataGradient(const Array<T> &incoming_gradient,
+                           const Array<T> &original_signal,
+                           const Array<accT> &original_filter,
+                           const Array<T> &convolved_output, af::dim4 stride,
+                           af::dim4 padding, af::dim4 dilation);
+
+template <typename T, typename accT>
+Array<T> conv2FilterGradient(const Array<T> &incoming_gradient,
+                             const Array<T> &original_signal,
+                             const Array<accT> &original_filter,
+                             const Array<T> &convolved_output, af::dim4 stride,
+                             af::dim4 padding, af::dim4 dilation);
+}
```

Comment:
Minor thing, and just for consistency, but can we add back `// namespace cpu` here?